server/utils/licenceStatus.js

```diff
@@ -0,0 +1,324 @@
+const {taskStates} = require('../models/taskStates');
+const {licenceStages} = require('../models/licenceStages');
+
+const {matcher, isPresent, isNotPresent, equalTo} = require('./jsonUtils');
+
+module.exports = {getLicenceStatus};
+
+function getLicenceStatus(licenceRecord) {
+
+    if (!licenceRecord || !licenceRecord.licence || !licenceRecord.status) {
+        return {
+            stage: licenceStages.UNSTARTED,
+            decisions: {},
+            tasks: {
+                exclusion: taskStates.UNSTARTED,
+                crdTime: taskStates.UNSTARTED,
+                suitability: taskStates.UNSTARTED,
+                eligibility: taskStates.UNSTARTED,
+                optOut: taskStates.UNSTARTED,
+                bassReferral: taskStates.UNSTARTED,
+                curfewAddress: taskStates.UNSTARTED,
+                riskManagement: taskStates.UNSTARTED,
+                curfewAddressReview: taskStates.UNSTARTED,
+                curfewHours: taskStates.UNSTARTED,
+                reportingInstructions: taskStates.UNSTARTED,
+                licenceConditions: taskStates.UNSTARTED,
+                approval: taskStates.UNSTARTED
+            }
+        };
+    }
+    const stage = licenceRecord.status;
+    const data = matcher(licenceRecord.licence);
+
+    const results = getRequiredState(stage, data);
+    const licenceStatus = results.reduce(combiner, {stage, decisions: {}, tasks: {}});
+
+    console.log(licenceStatus);
+    return licenceStatus;
+}
+
+function getRequiredState(stage, data) {
+
+    switch (stage) {
+        case licenceStages.DECIDED:
+        // fall through
+        case licenceStages.APPROVAL:
+            return [getEligibilityStageState(data), getRoStageState(data), getApprovalStageState(data)];
+        case licenceStages.PROCESSING_CA:
+        // fall through
+        case licenceStages.PROCESSING_RO:
+            return [getEligibilityStageState(data), getRoStageState(data)];
+        case licenceStages.ELIGIBILITY:
+            return [getEligibilityStageState(data)];
+        default:
+            return [];
+    }
+}
+
+const combiner = (acc, data) => {
+    Object.assign(acc.tasks, {...data.tasks});
+    Object.assign(acc.decisions, {...data.decisions});
```

Comment:
{...data.tasks} is equivalent to data.tasks

Reply:
Do you think this is more readable? Seems more in-keeping with what we do elsewhere
`const combiner = (acc, data) => {
    const combinedTasks = {...acc.tasks, ...data.tasks};
    const combinedDecisions = {...acc.decisions, ...data.decisions};

    return {
        ...acc,
        tasks: combinedTasks,
        decisions: combinedDecisions
    };
};
`

Reply:
I had something like that a one point. Have applied this change